lib/puppet_x/puppetlabs/cd4pe_job_helper.rb

```diff
@@ -0,0 +1,163 @@
+      @local_repo_dir = File.join(@working_dir, "cd4pe_job", "repo")
+      
+      @docker_run_args = docker_run_args.nil? ? '' : docker_run_args.join(' ')
+    end
+
+    def set_job_env_vars(task_params)
+      ENV["WEB_UI_ENDPOINT"] = task_params['cd4pe_web_ui_endpoint']
+      ENV["JOB_TOKEN"] = task_params['cd4pe_token']
+      ENV["JOB_OWNER"] = task_params['cd4pe_job_owner']
+      ENV["JOB_INSTANCE_ID"] = task_params['job_instance_id']
+    
+      user_specified_env_vars = task_params['env_vars']
+      if (!user_specified_env_vars.nil?)
+        user_specified_env_vars.each do |var|
+          pair = var.split("=")
+          key = pair[0]
+          value = pair[1]
+          ENV[key] = value
+      end
+    end
+
+    def make_working_dir(working_dir)
+      Dir.mkdir(working_dir) unless File.exists?(working_dir)
+    end
+
+    def get_job_script_and_control_repo
+      target_file = File.join(@working_dir, "cd4pe_job.tar.gz")
+      client = PuppetX::Puppetlabs::CD4PEClient.new
+      response = client.get_job_script_and_control_repo(target_file)
+      case response
+      when Net::HTTPNotFound
+        raise "Message: #{response.body}\nCode: #{response.code}"
+      when Net::HTTPServerError
+        raise "Unknown HTTP Error with code: #{response.code} and body #{response.body}"
+      end
+
+      target_file
+    end
+
+    def run_job
+      result = execute_manifest(MANIFEST_TYPE[:JOB])
+      
+      if (result[:exit_code] == 0)
+        on_job_success(result)
+      else
+        on_job_failure(result)
+      end
+    end
+
+    def on_job_success(result)
+      exit_code = 0
+    
+      output = {}
+      output[:job] = {
+        exit_code: result[:exit_code], 
+        message: result[:message]
+      }
+    
+      # if a AFTER_JOB_SUCCESS script exists, run it now!
+      run_after_success = File.exists?(File.join(@local_jobs_dir, MANIFEST_TYPE[:AFTER_JOB_SUCCESS]))
+      if (run_after_success)
+        success_script_result = execute_manifest(MANIFEST_TYPE[:AFTER_JOB_SUCCESS])
+        exit_code = success_script_result[:exit_code]
+        output[:after_job_success] = {
+          exit_code: exit_code, 
+          message: success_script_result[:message]
+        }
+      end
+    
+      on_script_complete(output, exit_code)
+    end
+
+    def on_job_failure(result)
+      output = {}
+      output[:job] = {
+        exit_code: result[:exit_code], 
+        message: result[:message]
+      }
+    
+      # if a AFTER_JOB_FAILURE script exists, run it now!
+      run_after_failure = File.exists?(File.join(@local_jobs_dir, MANIFEST_TYPE[:AFTER_JOB_FAILURE]))
+      if (run_after_failure)
+        failure_script_result = execute_manifest(MANIFEST_TYPE[:AFTER_JOB_FAILURE])
+        output[:after_job_failure] = {
+          exit_code: failure_script_result[:exit_code], 
+          message: failure_script_result[:message]
+        }
+      end
+
+      on_script_complete(output, 1)
```

Comment:
why do we exit with hardcoded exit status vs what job returned?

Reply:
So, two things here:

1. This code is outdated. Rather than exiting here, we instead return "output" to the caller to handle themselves.
2. That being said, for our purposes, we still end up exiting 1 regardless of the actual exit code. This is because we are actually running _multiple_ scripts with each job, and this exit code represents the overall code we want Bolt to act upon. We do, however, return the actual exit code from each script in the json output that we print for Bolt such that the caller can still use that info if they want to (see the log output above for an example of how we capture multiple error codes).

Fun fact: Bolt prefers we print a single JSON object to STDOUT when running a bolt task in order for it to capture it properly: https://puppet.com/docs/bolt/latest/writing_tasks.html#returning-structured-output